File: pacman-reloaded/src/main/java/components/pacman.rb
require 'java'

java_import Java::extensions.shapes.Shapeable
require_relative './mixins/movable'
require_relative './mixins/handleable'
class Pacman < BasicPacmanComponent
  include Movable
  include Handleable

  def initialize(shape, appearance, x, y, x_vector, y_vector, speed)
    super shape, appearance, x, y
    initialize_vector x_vector, y_vector, speed
  end

  def update(delta_state)
    update_vector delta_state, uVector
    if can_move? delta_state
      move(delta_state)
    end
    super
  end

  def can_move?(delta_state)
    x, y = shape.x, shape.y
    new_x = calculate_position x, uVector.x, speed_factor(delta_state)
    new_y = calculate_position y, uVector.y, speed_factor(delta_state)
    scene.can_occupy?(new_x, new_y, shape)
  end

  java_signature 'double getX()'

  def getX
    super
  end

  java_signature 'double getY()'

  def getY
    super
  end
end

File: pacman-reloaded/src/main/java/components/Pacman.java
package components;


import com.uqbar.vainilla.DeltaState;
import com.uqbar.vainilla.UnitVector2D;
import com.uqbar.vainilla.appearances.Appearance;
import com.uqbar.vainilla.events.constants.Key;
import extensions.shapes.Circle;

public class Pacman extends PacmanComponent{

    public Pacman(Appearance appearance, double xPos, double yPos, double xVec, double yVec, double speei1, Circle shape) {
        super(appearance, xPos, yPos, xVec, yVec, speei1, shape);
    }

    public void update(DeltaState delta_state) {
        updateVector(delta_state, getUVector());
        if (can_move(delta_state))
            super.update(delta_state);
    }

    public void updateVector(DeltaState delta_state, UnitVector2D vector) {
        if (delta_state.isKeyPressed(Key.UP)) {
            vector.setY(-1);
            vector.setX(0);
        }
        if (delta_state.isKeyPressed(Key.DOWN)) {
            vector.setY(1);
            vector.setX(0);
        }
        if (delta_state.isKeyPressed(Key.RIGHT)) {
            vector.setX(1);
            vector.setY(0);
        }
        if (delta_state.isKeyPressed(Key.LEFT)) {
            vector.setX(-1);
            vector.setY(0);
        }

    }

}
